### PR TITLE
Add character strategies

### PR DESCRIPTION
**Pull request by brandoncc**
@ThePrimeagen What do you think about the way I implemented styled text in this PR?

I also included an `.editorconfig` so those of us with soft tabs set to two spaces don't break your code style 😃

### CODE REVIEW

**Comment by brandoncc** [comment]
@ThePrimeagen I missed the fact that only the "NORMAL" mode should be colored. I added switching strategies in my last two commits.

Is it just my system, or does it seem like "bold" isn't working?

**Comment by ThePrimeagen** [comment]
I want them all to be colored, but a switch mode should allow for this.  

Let me review this.

**Comment by ThePrimeagen** [comment]
*TL/DR;* Big ups buddy.  I really like this.  Seems like an awesome change.  Read the `1` section about garbage, address changes, and i'll merge.  Great work here.

2 things:  first definitely required before merge, second just me pontificating, join in and provide some thoughts.

#### 1. Garbage.  

JavaScript really sucks creating tons of garbage.  Cache your results.  Try to come up with some strategies that don't involve creating so many recreated pieces of data.  I'll make them with a `1.` comment.  Effectively, why recompute these changes every 16ms (our current tick rate for the client).

#### 2. PositionComponent in general.  

The position component is slowly turning into a dumpster of all things related to rendering. As of now, I am fine with where we are at, but I do want to consider a better entity interface.  What I am thinking of is separating out the `PositionComponent` from a `RenderInfoComponent` that contains a bit more in depth info. 

My basic idea behind this will lead to hopefully a more robust Networking solution.  That way we could consider a better "encoding" strategy for Network communications.  The network communications could pipe out messages per changeset of properties on *Component.  This    

##### Don't change anything right now

I want to think more about this.  I feel something great coming, but I am also trying not to refactor to make "nicer" code to work with.  I will refactor if I am positive I can significantly reduce complexity as each new feature gets implemented.

**Comment by brandoncc** [comment]
I'm unsure whether re-rendering the chars autistically after changing the strategy is the right call. Maybe the code which changes the strategy should have to call setChar again?

File: src/characters/styled.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    fg?: string;
+    bg?: string;
+};
+
+type StyleTags = [string, string];
+
+const decorators = {
+    bold(styleTags: StyleTags, enabled: boolean) {
+        // only add tags if enabled is `true`
+        if (enabled) { buildTag(styleTags, 'bold'); }
```
Comment (ThePrimeagen) [inline src/characters/styled.ts]: nit.  general style, no one liners ;)

**Comment by brandoncc** [comment]
Apparently my phone autocorrected "automatically" to "autistically"...

File: src/characters/styled.ts (hunk 1 above)
Comment (brandoncc) [reply src/characters/styled.ts]: I wasn't sure what your preference was on this. I tried my best to copy any preferences I could deduce from reading your code, thanks for the pointer.

**Comment by brandoncc** [comment]
There is one thing I am unsure of. I cached `_rawChar` so that I could call `setChar` within `setCharacterStrategy`. In the case of `Mode`, this actually has the effect of double rendering, since we also always call `setChar` along-side `setCharacterStrategy`. I'm wondering if `setCharacterStrategy` should not call `setChar`, and the code which changes the strategy should also have to call `setChar` if it wants to re-render?

---

As for #2 above, I 100% agree with you. My first two implementations of this feature were (1) inheritance and (2) continuing down the path of your styled component.

I could feel the coupling tightening around my neck with every line of code I wrote in both situations.

I had multiple situations where I thought about refactoring even my own code to make it nicer to deal with, or make some (possibly premature) optimizations...I refrained.

At the very least, I think the name `Position` has been outgrown, but I definitely feel like the class has multiple responsibilities now. It should probably be split out to rely on a completely separate renderer. I think the character strategies are a good first step toward doing that.
